storage/innobase/include/arch0arch.h

```diff
@@ -0,0 +1,1984 @@
+  ARCH_STATE_ACTIVE -right-> ARCH_STATE_PREPARE_IDLE : Stop archiving
+  ARCH_STATE_PREPARE_IDLE -right-> ARCH_STATE_IDLE : All data archived
+  ARCH_STATE_IDLE -down-> ARCH_STATE_ABORT : Shutdown or Fatal Error
+  ARCH_STATE_PREPARE_IDLE --> ARCH_STATE_ACTIVE : Resume archiving
+  ARCH_STATE_IDLE --> ARCH_STATE_ACTIVE : Start archiving
+  ARCH_STATE_ABORT -down-> [*]
+
+@enduml */
+enum Arch_State
+{
+  /** Archiver is initialized */
+  ARCH_STATE_INIT = 0,
+
+  /** Archiver is active and archiving data */
+  ARCH_STATE_ACTIVE,
+
+  /** Archiver is processing last data chunks before idle state */
+  ARCH_STATE_PREPARE_IDLE,
+
+  /** Archiver is idle */
+  ARCH_STATE_IDLE,
+
+  /** Server is in read only mode, and hence the archiver */
+  ARCH_STATE_READ_ONLY,
+
+  /** Archiver is aborted */
+  ARCH_STATE_ABORT
+};
+
+/** Archived data block state.
+A data block is a block in memory that holds dirty page IDs before persisting
+into disk. Shown below is the state transfer diagram for a data block.
+
+@startuml
+
+  state ARCH_BLOCK_INIT
+  state ARCH_BLOCK_ACTIVE
+  state ARCH_BLOCK_READY_TO_FLUSH
+  state ARCH_BLOCK_FLUSHED
+
+  [*] -down-> ARCH_BLOCK_INIT
+  ARCH_BLOCK_INIT -> ARCH_BLOCK_ACTIVE : Writing page ID
+  ARCH_BLOCK_ACTIVE -> ARCH_BLOCK_READY_TO_FLUSH : Block is full
+  ARCH_BLOCK_READY_TO_FLUSH -> ARCH_BLOCK_FLUSHED : Block is flushed
+  ARCH_BLOCK_FLUSHED --> ARCH_BLOCK_ACTIVE : Writing page ID
+  ARCH_BLOCK_FLUSHED -down-> [*]
+
+@enduml */
+enum Arch_Blk_State
+{
+  /** Data block is initialized */
+  ARCH_BLOCK_INIT = 0,
+
+  /** Data block is active and having data */
+  ARCH_BLOCK_ACTIVE,
+
+  /** Data block is full but not flushed to disk */
+  ARCH_BLOCK_READY_TO_FLUSH,
+
+  /** Data block is flushed and can be reused */
+  ARCH_BLOCK_FLUSHED
+};
+
+/** Archiver block type */
+enum Arch_Blk_Type
+{
+  /* Block which holds reset information */
+  ARCH_RESET_BLOCK = 0,
+
+  /* Block which holds archived page IDs */
+  ARCH_DATA_BLOCK
+};
+
+/** Archiver block flush type */
+enum Arch_Blk_Flush_Type
+{
+  /** Flush when block is full */
+  ARCH_FLUSH_NORMAL = 0,
+
+  /** Flush partial block.
+  Needed for persistent page tracking. */
+  ARCH_FLUSH_PARTIAL
+};
+
+/** Page Archive doublewrite buffer block offsets */
+enum Arch_Page_Dblwr_Offset
+{
+  /** Archive doublewrite buffer page offset for RESET page. */
+  ARCH_PAGE_DBLWR_RESET_PAGE = 0,
+
+  /* Archive doublewrite buffer page offset for FULL FLUSH page. */
+  ARCH_PAGE_DBLWR_FULL_FLUSH_PAGE,
+
+  /* Archive doublewrite buffer page offset for PARTIAL FLUSH page. */
+  ARCH_PAGE_DBLWR_PARTIAL_FLUSH_PAGE
+};
+
+/** Forward declarations */
+class Arch_Group;
+class Arch_Log_Sys;
+class Arch_Dblwr_Ctx;
+class Arch_Recv_Group_Info;
+
+/** Guard to release resources safely */
+class Arch_scope_guard
+{
+ public:
+  /** Attach a function to the guard which releases some resource. */
+  Arch_scope_guard(std::function<void()> function) { m_cleanup = function; }
```

Comment:
This seems to duplicate `make_scope_exit`. It is better to use common code. Besides, it looks like this is an unreachable class.